File: test/csharpxml/Test5.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml.Serialization;

namespace csharpxml.Test5
{
    [XmlRoot(ElementName ="node")]
    public class DBusNode : XmlTestObject<DBusNode>
    {
        [XmlAttribute]
        public string name;
        [XmlElement(ElementName = "interface")]
        public DBusInterface[] Interfaces;

        public override void PopulateTestObject(int subtest)
        {
            switch (subtest)
            {
                case 0:
                    break;
                case 1:
                    var face = new DBusInterface();
                    face.name = "MyInterface";
                    var m = new DBusMethod();
                    m.name = "MyMethod";
                    m.Args = new DBusArg[]
                    {
                        new DBusArg("Arg1", "s", "in"),
                        new DBusArg("Out1", "s", "out"),
                    };
                    face.Methods = new DBusMethod[] { m };
                    var s = new DBusSignal();
                    s.name = "MySig";
                    s.Args = new DBusArg[]
                    {
                        new DBusArg("SigArg", "si", null),
                    };
                    face.Signals = new DBusSignal[] { s };
                    face.Properties = new DBusProperty[]
                    {
                        new DBusProperty("MyProp", "s", "readwrite"),
                    };
                    Interfaces = new DBusInterface[] { face };
                    break;
            }
        }

    }

    public class DBusArg
    {
        [XmlAttribute]
        public string name;
        [XmlAttribute]
        public string type;
        [XmlAttribute]
        public string direction;

        public DBusArg() { }
        public DBusArg(string n, string t, string d) { name = n; type = t; direction = d; }
    }

    public class DBusProperty
    {
        [XmlAttribute]
        public string name;
        [XmlAttribute]
        public string type;
        [XmlAttribute]
        public string access;

        public DBusProperty() { }
        public DBusProperty(string n, string t, string a) { name = n; type = t; access = a; }
    }

    public class DBusMethod
    {
        [XmlAttribute]
        public string name;
        [XmlElement(ElementName = "arg")]
        public DBusArg[] Args;

        public DBusMethod() { }
    }

    public class DBusSignal
    {
        [XmlAttribute]
        public string name;
        [XmlElement(ElementName = "arg")]
        public DBusArg[] Args;

        public DBusSignal() { }
    }

    public class DBusInterface
    {
        [XmlAttribute]
        public string name;
        [XmlElement(ElementName = "method")]
        public DBusMethod[] Methods;
        [XmlElement(ElementName = "signal")]
        public DBusSignal[] Signals;
        [XmlElement(ElementName = "property")]
        public DBusProperty[] Properties;

        public DBusInterface() { }
    }

}

File: test/Test5.js
'use strict'
const xml = require('../xml-csharp-cereal');
const Long = require("long"); // to test simple decode/encode overrides

// Try serializing some standard form of XML like DBus introspection

module.exports.GetFactory = function()
{
    return (new xml.XmlTemplateFactory(DBusNode, DBusInterface, DBusSignal, DBusMethod, DBusProperty, DBusArg));
}

class DBusNode
{
    constructor()
    {
        this.name = null; // string
        this.Interfaces = null; // array of DBusInterface
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this, null, 'node');
        temp.addString('name').attr();
        temp.add('Interfaces', 'DBusInterface', ['interface']).flatArr();
        return temp;
    }
}
module.exports.TestClass=DBusNode;

class DBusArg
{
    constructor()
    {
        this.name = null;
        this.type = null;
        this.direction = null;
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addString('name').attr();
        temp.addString('type').attr();
        temp.addString('direction').attr();
        return temp;
    }
}

class DBusProperty
{
    constructor()
    {
        this.name = null;
        this.type = null;
        this.access = null;
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addString('name').attr();
        temp.addString('type').attr();
        temp.addString('access').attr();
        return temp;
    }
}

class DBusMethod
{
    constructor()
    {
        this.name = null;
        this.Args = null;
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addString('name').attr();
        temp.add('Args', 'DBusArg', ['arg']).flatArr();
        return temp;
    }
}

class DBusSignal
{
    constructor()
    {
        this.name = null;
        this.Args = null;
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addString('name').attr();
        temp.add('Args', 'DBusArg', ['arg']).flatArr();
        return temp;
    }
}

class DBusInterface
{
    constructor()
    {
        this.name = null;
        this.Methods = null;
        this.Signals = null;
        this.Properties = null;
    }
    static getXmlTemplate()
    {
        var temp = new xml.XmlTemplate(this);
        temp.addString('name').attr();
        temp.add('Methods', 'DBusMethod', ['method']).flatArr();
        temp.add('Signals', 'DBusSignal', ['signal']).flatArr();
        temp.add('Properties', 'DBusProperty', ['property']).flatArr();
        return temp;
    }
}
